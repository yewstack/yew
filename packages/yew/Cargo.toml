[package]
name = "yew"
version = "0.17.4"
edition = "2018"
authors = [
  "Denis Kolodin <deniskolodin@gmail.com>",
  "Justin Starry <justin@yew.rs>",
]
repository = "https://github.com/yewstack/yew"
homepage = "https://github.com/yewstack/yew"
documentation = "https://docs.rs/yew/"
license = "MIT/Apache-2.0"
readme = "../README.md"
keywords = ["web", "webasm", "javascript"]
categories = ["gui", "wasm", "web-programming"]
description = "A framework for making client-side single-page apps"

[dependencies]
anyhow = "1"
anymap = "0.12"
console_error_panic_hook = "0.1"
gloo = "0.2.1"
http = "0.2"
indexmap = { version = "1.5", features = ["std"] }
js-sys = "0.3"
log = "0.4"
serde = { version = "1.0", features = ["derive"] }
serde_json = "1.0"
slab = "0.4"
thiserror = "1"
wasm-bindgen = "0.2.60"
wasm-bindgen-futures = "0.4"
yew-macro = { version = "0.17.0", path = "../yew-macro" }

# optional encodings
bincode = { version = "1", optional = true }
rmp-serde = { version = "0.15.0", optional = true }
serde_cbor = { version = "0.11.1", optional = true }
serde_yaml = { version = "0.8.4", optional = true }
toml = { version = "0.5", optional = true }

[dependencies.web-sys]
version = "0.3"
features = [
  "AbortController",
  "AbortSignal",
  "AnimationEvent",
  "BinaryType",
  "Blob",
  "BlobPropertyBag",
  "console",
  "DedicatedWorkerGlobalScope",
  "Document",
  "DomTokenList",
  "DragEvent",
  "Element",
  "ErrorEvent",
  "Event",
  "EventTarget",
  "File",
  "FileList",
  "FileReader",
  "FocusEvent",
  "Headers",
  "HtmlElement",
  "HtmlButtonElement",
  "HtmlInputElement",
  "HtmlSelectElement",
  "HtmlTextAreaElement",
  "InputEvent",
  "KeyboardEvent",
  "Location",
  "MessageEvent",
  "MouseEvent",
  "Node",
  "ObserverCallback",
  "PointerEvent",
  "ProgressEvent",
  "ReferrerPolicy",
  "Request",
  "RequestCache",
  "RequestCredentials",
  "RequestInit",
  "RequestMode",
  "RequestRedirect",
  "Response",
  "Storage",
  "Text",
  "TouchEvent",
  "TransitionEvent",
  "UiEvent",
  "Url",
  "WebSocket",
  "WheelEvent",
  "Window",
  "Worker",
  "WorkerGlobalScope",
  "WorkerOptions",
]

[dev-dependencies]
base64 = "0.13.0"
bincode = "1"
easybench-wasm = "0.2.1"
rmp-serde = "0.15.0"
rustversion = "1.0"
serde_derive = "1"
ssri = "6.0.0"
trybuild = "1.0"
wasm-bindgen-test = "0.3.4"

[features]
default = ["agent"]
doc_test = []
wasm_test = []
wasm_bench = []
agent = ["bincode"]
yaml = ["serde_yaml"]
msgpack = ["rmp-serde"]
cbor = ["serde_cbor"]

[package.metadata.docs.rs]
features = ["yaml", "cbor", "toml", "msgpack", "doc_test"]
