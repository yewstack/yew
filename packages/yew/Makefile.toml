[env]
CARGO_MAKE_CLIPPY_ARGS = "--features cbor,msgpack,toml,yaml -- --deny=warnings"

[tasks.test]
extend = "core::wasm-pack-base"
command = "wasm-pack"
args = [
    "test",
    "@@split(YEW_TEST_FLAGS, )",
    "--",
    "--features",
    "${YEW_TEST_FEATURES}",
]

[tasks.doc-test]
clear = true
run_task = { name = ["doc-test-normal"], fork = true }

[tasks.doc-test-normal]
command = "cargo"
args = [
    "test",
    "--doc",
    "--features",
    "doc_test,wasm_test,yaml,msgpack,cbor,toml",
]

[tasks.bench]
extend = "core::wasm-pack-base"
command = "wasm-pack"
args = [
    "test",
    "--release",
    "--firefox",
    "--headless",
    "--",
    "--features",
    "wasm_bench",
    "bench",
]
