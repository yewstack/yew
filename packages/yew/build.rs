use std::env;

pub fn main() {
    let using_web_sys = cfg!(feature = "web_sys");
    let using_std_web = cfg!(feature = "std_web");
    if using_web_sys && using_std_web {
        panic!("Yew does not allow the `web_sys` and `std_web` cargo features to be used simultaneously");
    } else if !using_web_sys && !using_std_web {
        panic!("Yew requires selecting either the `web_sys` or `std_web` cargo feature");
    }

    let using_cargo_web = env::var("COMPILING_UNDER_CARGO_WEB").is_ok();
    if using_cargo_web && using_web_sys {
        panic!("cargo-web is not compatible with web-sys");
    }
}
