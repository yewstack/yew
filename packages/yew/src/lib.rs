#![allow(clippy::needless_doctest_main)]
#![doc(html_logo_url = "https://static.yew.rs/logo.svg")]

//! # Yew Framework - API Documentation
//!
//! Yew is a modern Rust framework for creating multi-threaded front-end web apps using WebAssembly
//!
//! - Features a macro for declaring interactive HTML with Rust expressions. Developers who have experience using JSX in React should feel quite at home when using Yew.
//! - Achieves high performance by minimizing DOM API calls for each page render and by making it easy to offload processing to background web workers.
//! - Supports JavaScript interoperability, allowing developers to leverage NPM packages and integrate with existing JavaScript applications.
//!
//! ### Supported Targets
//! - `wasm32-unknown-unknown`
//!
//! ### Important Notes
//! - Yew is not (yet) production ready but is great for side projects and internal tools
#![cfg_attr(
    feature = "web_sys",
    doc = " - If your app is built with `stdweb`, we recommend using [`yew-stdweb`](https://docs.rs/yew-stdweb) instead."
)]
//!
//! ## Example
//!
//! ```rust
//! use yew::prelude::*;
//!
//! enum Msg {
//!     AddOne,
//! }
//!
//! struct Model {
//!     link: ComponentLink<Self>,
//!     value: i64,
//! }
//!
//! impl Component for Model {
//!     type Message = Msg;
//!     type Properties = ();
//!
//!     fn create(_props: Self::Properties, link: ComponentLink<Self>) -> Self {
//!         Self {
//!             link,
//!             value: 0,
//!         }
//!     }
//!
//!     fn update(&mut self, msg: Self::Message) -> ShouldRender {
//!         match msg {
//!             Msg::AddOne => {
//!                 self.value += 1;
//!                 true
//!             }
//!         }
//!     }
//!
//!     fn change(&mut self, _props: Self::Properties) -> ShouldRender {
//!         false
//!     }
//!
//!     fn view(&self) -> Html {
//!         html! {
//!             <div>
//!                 <button onclick=self.link.callback(|_| Msg::AddOne)>{ "+1" }</button>
//!                 <p>{ self.value }</p>
//!             </div>
//!         }
//!     }
//! }
//!
//!# fn dont_execute() {
//! fn main() {
//!     yew::start_app::<Model>();
//! }
//!# }
//! ```
//!

#![deny(
    missing_docs,
    missing_debug_implementations,
    bare_trait_objects,
    anonymous_parameters,
    elided_lifetimes_in_paths
)]
#![allow(macro_expanded_macro_exports_accessed_by_absolute_paths)]
#![recursion_limit = "512"]
extern crate self as yew;

/// This macro provides a convenient way to create [`Classes`].
///
/// The macro takes a list of items similar to the [`vec!`] macro and returns a [`Classes`] instance.
/// Each item can be of any type that implements `Into<Classes>` (See the implementations on [`Classes`] to learn what types can be used).
///
/// # Example
///
/// ```
/// # use yew::prelude::*;
/// # fn test() {
/// let conditional_class = Some("my-other-class");
/// let vec_of_classes = vec!["one-bean", "two-beans", "three-beans", "a-very-small-casserole"];
///
/// html! {
///     <div class=classes!("my-container-class", conditional_class, vec_of_classes)>
///         // ...
///     </div>
/// };
/// # }
/// ```
pub use yew_macro::classes;

/// This macro implements JSX-like templates.
///
/// This macro always returns [`Html`].
/// If you need to preserve the type of a component, use the [`html_nested!`] macro instead.
///
/// More information about using the `html!` macro can be found in the [Yew Docs]
///
/// [`Html`]: ./html/type.Html.html
/// [`html_nested!`]: ./macro.html_nested.html
/// [Yew Docs]: https://yew.rs/docs/en/concepts/html/
pub use yew_macro::html;

/// This macro is similar to [`html!`], but preserves the component type instead
/// of wrapping it in [`Html`].
///
/// That macro is useful when, for example, in a typical implementation of a list
/// component (let's assume it's called `List`).
/// In a typical implementation you might find two component types -- `List` and `ListItem`.
/// Only `ListItem` components are allowed to be children of List`.
///
/// You can find an example implementation of this in the [`nested_list`] example.
/// That example shows, how to create static lists with their children.
///
/// ```
/// # use yew::prelude::*;
/// use yew::html::ChildrenRenderer;
/// use yew::virtual_dom::VChild;
///
/// #[derive(Clone, Properties)]
/// struct List {
///   children: ChildrenRenderer<ListItem>,
/// }
/// impl Component for List {
/// #   type Message = ();
///   type Properties = Self;
///   // ...
/// #   fn create(props: Self::Properties, _: ComponentLink<Self>) -> Self { props }
/// #   fn update(&mut self, _: Self::Message) -> ShouldRender { false }
/// #   fn change(&mut self, _: Self::Properties) -> ShouldRender { false }
/// #   fn view(&self) -> Html { unimplemented!() }
/// }
///
/// #[derive(Clone)]
/// struct ListItem;
/// impl Component for ListItem {
/// #   type Message = ();
/// #   type Properties = ();
///   // ...
/// #   fn create(_: Self::Properties, _: ComponentLink<Self>) -> Self { Self }
/// #   fn update(&mut self, _: Self::Message) -> ShouldRender { false }
/// #   fn change(&mut self, _: Self::Properties) -> ShouldRender { false }
/// #   fn view(&self) -> Html { unimplemented!() }
/// }
///
/// // Required for ChildrenRenderer
/// impl From<VChild<ListItem>> for ListItem {
///   fn from(child: VChild<ListItem>) -> Self { Self }
/// }
///
/// impl Into<Html> for ListItem {
///   fn into(self) -> Html { self.view() }
/// }
///
/// // You can use `List` with nested `ListItem` components.
/// // Using any other kind of element would result in a compile error.
/// # fn test() -> Html {
/// html! {
///   <List>
///     <ListItem/>
///     <ListItem/>
///     <ListItem/>
///   </List>
/// }
/// # }
/// # fn test_iter() -> Html {
/// # let some_iter = (0..10);
/// // In many cases you might want to create the content dynamically.
/// // To do this, you can use the following code:
/// html! {
///   <List>
///     { for some_iter.map(|_| html_nested!{ <ListItem/> }) }
///   </List>
/// }
/// # }
/// ```
///
/// If you used the [`html!`] macro instead of `html_nested!`, the code would
/// not compile because we explicitly indicated to the compiler that `List`
/// can only contain elements of type `ListItem` using [`ChildrenRenderer<ListItem>`],
/// while [`html!`] creates items of type [`Html`].
///
///
/// [`html!`]: ./macro.html.html
/// [`Html`]: ./html/type.Html.html
/// [`nested_list`]: https://github.com/yewstack/yew/tree/master/examples/nested_list
/// [`ChildrenRenderer<ListItem>`]: ./html/struct.ChildrenRenderer.html
pub use yew_macro::html_nested;

/// Build [`Properties`] outside of the [`html!`] macro.
///
/// It's already possible to create properties like normal Rust structs
/// but if there are lots of optional props the end result is often needlessly verbose.
/// This macro allows you to build properties the same way the [`html!`] macro does.
///
/// The macro doesn't support special props like `ref` and `key`, they need to be set in the [`html!`] macro.
///
/// You can read more about `Properties` in the [Yew Docs].
///
/// # Example
///
/// ```
/// # use yew::prelude::*;
/// use std::borrow::Cow;
///
/// #[derive(Clone, Properties)]
/// struct Props {
///     #[prop_or_default]
///     id: usize,
///     name: Cow<'static, str>,
/// }
///
/// struct Model(Props);
/// impl Component for Model {
/// #   type Message = ();
///     type Properties = Props;
///     // ...
/// #   fn create(_props: Self::Properties, _link: ComponentLink<Self>) -> Self { unimplemented!() }
/// #   fn update(&mut self, _msg: Self::Message) -> ShouldRender { unimplemented!() }
/// #   fn change(&mut self, _props: Self::Properties) -> ShouldRender { unimplemented!() }
/// #   fn view(&self) -> Html { unimplemented!() }
/// }
///
/// # fn foo() -> Html {
/// // You can build props directly ...
/// let props = yew::props!(Props { name: Cow::from("Minka") });
/// # assert_eq!(props.name, "Minka");
/// // ... or build the associated properties of a component
/// let props = yew::props!(Model::Properties { id: 2, name: Cow::from("Lemmy") });
/// # assert_eq!(props.id, 2);
///
/// // Use the `with props` syntax to create a component with the props.
/// html! {
///     <Model key=1 with props />
/// }
/// # }
/// ```
///
/// [`html!`]: ./macro.html.html
/// [`Properties`]: ./html/trait.Properties.html
/// [yew docs]: https://yew.rs/docs/en/concepts/components/properties
pub use yew_macro::props;

/// This module contains macros which implements html! macro and JSX-like templates
pub mod macros {
    pub use crate::classes;
    pub use crate::html;
    pub use crate::html_nested;
    pub use crate::props;
}

pub mod app;
pub mod callback;
pub mod format;
pub mod html;
mod scheduler;
pub mod utils;
pub mod virtual_dom;

#[cfg(feature = "agent")]
pub mod agent;

pub use web_sys;

/// The module that contains all events available in the framework.
pub mod events {
    pub use crate::html::{ChangeData, InputData};

    #[doc(no_inline)]
    pub use web_sys::{
        AnimationEvent, DragEvent, ErrorEvent, Event, FocusEvent, InputEvent, KeyboardEvent,
        MouseEvent, PointerEvent, ProgressEvent, TouchEvent, TransitionEvent, UiEvent, WheelEvent,
    };
}

/// Starts an app mounted to a body of the document.
pub fn start_app<COMP>()
where
    COMP: Component,
    COMP::Properties: Default,
{
    App::<COMP>::new().mount_to_body();
}

/// Starts an app mounted to a body of the document.
pub fn start_app_with_props<COMP>(props: COMP::Properties)
where
    COMP: Component,
{
    App::<COMP>::new().mount_to_body_with_props(props);
}

/// The Yew Prelude
///
/// The purpose of this module is to alleviate imports of many common types:
///
/// ```
/// # #![allow(unused_imports)]
/// use yew::prelude::*;
/// ```
pub mod prelude {
    #[cfg(feature = "agent")]
    pub use crate::agent::{Bridge, Bridged, Dispatched, Threaded};
    pub use crate::app::App;
    pub use crate::callback::Callback;
    pub use crate::events::*;
    pub use crate::html::{
        Children, ChildrenWithProps, Classes, Component, ComponentLink, Html, NodeRef, Properties,
        ShouldRender,
    };
    pub use crate::macros::{classes, html, html_nested};

    /// Prelude module for creating worker.
    #[cfg(feature = "agent")]
    pub mod worker {
        pub use crate::agent::{
            Agent, AgentLink, Bridge, Bridged, Context, HandlerId, Job, Private, Public,
        };
    }
}

pub use self::prelude::*;
