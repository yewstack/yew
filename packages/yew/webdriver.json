{
  "goog:chromeOptions": {
    "args": [
      "--use-fake-ui-for-media-stream",
      "--use-fake-device-for-media-stream"
    ]
  }
}
