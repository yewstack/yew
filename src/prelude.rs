//! The Yew Prelude
//!
//! The purpose of this module is to alleviate imports of many common types:
//!
//! ```
//! # #![allow(unused_imports)]
//! use yew::prelude::*;
//! ```
extern crate stdweb;

pub use html::{
    Component,
    ComponentLink,
    Href,
    Html,
    ChangeData,
    InputData,
    Renderable,
    ShouldRender,
};

pub use app::App;

pub use callback::Callback;

pub use stdweb::web::event::{
    BlurEvent,
    ClickEvent,
    DoubleClickEvent,
    IKeyboardEvent,
    IMouseEvent,
    KeyDownEvent,
    KeyPressEvent,
    KeyUpEvent,
    MouseDownEvent,
    MouseMoveEvent,
    MouseOutEvent,
    MouseOverEvent,
    MouseUpEvent,
};

pub use agent::{
    Ambit,
    Bridge,
    Worker,
};

/// Prelude module for creating worker.
pub mod worker {
    pub use agent::{
        Addr,
        Agent,
        AgentLink,
        HandlerId,
        Transferable,
        Worker,
    };
}
