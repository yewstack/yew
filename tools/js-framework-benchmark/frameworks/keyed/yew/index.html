<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8">
    <title>Yew</title>
    <link href="/css/currentStyle.css" rel="stylesheet"/>
    <base href="bundled-dist/"></base>
  </head>
  <body>
    <div id='main'></div>
    <script type="module">
      import init from './js-framework-benchmark-yew.js';
      init('./js-framework-benchmark-yew_bg.wasm');
    </script>
  </body>
</html>
