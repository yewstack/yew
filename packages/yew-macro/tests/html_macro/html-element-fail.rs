use yew::prelude::*;

struct NotToString;

fn compile_fail() {
    html! { <div> };
    html! { <div><div> };
    html! { </div> };
    html! { <div><div></div> };
    html! { <div></div><div></div> };
    html! { <div></span> };
    html! { <tag-a></tag-b> };
    html! { <div></span></div> };
    html! { <img /></img> };
    html! { <div>Invalid</div> };

    html! { <input attr=1 attr=2 /> };
    html! { <input value="123" value="456" /> };
    html! { <input kind="checkbox" kind="submit" /> };
    html! { <input checked=true checked=false /> };
    html! { <input disabled=true disabled=false /> };
    html! { <option selected=true selected=false /> };
    html! { <div class="first" class="second" /> };

    html! { <input checked=1 /> };
    html! { <input disabled=1 /> };
    html! { <option selected=1 /> };
    html! { <input type=() /> };
    html! { <input value=() /> };
    html! { <a href=() /> };

    html! { <input onclick=1 /> };
    html! { <input onclick=Callback::from(|a: String| ()) /> };

    html! { <input string=NotToString /> };

    html! { <input ref=() /> };
    html! { <input ref=() ref=() /> };

    html! { <input type="text"></input> };
    html! { <iNpUt type="text"></iNpUt> };

    html! { <@></@> };
    html! { <@{"test"}></@{"test"}> };
    html! { <@{55}></@> };
    html! { <@/> };

    html! { <a media?="media" /> };
    html! { <a media?=Some(NotToString) /> };
    html! { <input disabled?=Some(true) /> };
    html! { <input type?="kind" /> };
    html! { <input type?=Some(NotToString) /> };
    html! { <li value?="value" /> };
    html! { <li value?=Some(NotToString) /> };
    html! { <a href?=Some(5) /> };
    html! { <a href?="href" /> };
    html! { <a href?=Some(NotToString) /> };
    html! { <input checked?=Some(false) /> };
    html! { <input ref?=() /> };
    html! { <input onfocus?=Some(5) /> };
    html! { <input onfocus?=Callback::from(|_| ()) /> };

    html! { <div class=("deprecated", "warning") /> };
}

fn main() {}
