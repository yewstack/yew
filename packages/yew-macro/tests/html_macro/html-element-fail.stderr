error: this opening tag has no corresponding closing tag
 --> $DIR/html-element-fail.rs:6:13
  |
6 |     html! { <div> };
  |             ^^^^^

error: this opening tag has no corresponding closing tag
 --> $DIR/html-element-fail.rs:7:18
  |
7 |     html! { <div><div> };
  |                  ^^^^^

error: this closing tag has no corresponding opening tag
 --> $DIR/html-element-fail.rs:8:13
  |
8 |     html! { </div> };
  |             ^^^^^^

error: this opening tag has no corresponding closing tag
 --> $DIR/html-element-fail.rs:9:13
  |
9 |     html! { <div><div></div> };
  |             ^^^^^

error: only one root html element is allowed (hint: you can wrap multiple html elements in a fragment `<></>`)
  --> $DIR/html-element-fail.rs:10:24
   |
10 |     html! { <div></div><div></div> };
   |                        ^^^^^^^^^^^

error: this closing tag has no corresponding opening tag
  --> $DIR/html-element-fail.rs:11:18
   |
11 |     html! { <div></span> };
   |                  ^^^^^^^

error: this closing tag has no corresponding opening tag
  --> $DIR/html-element-fail.rs:12:20
   |
12 |     html! { <tag-a></tag-b> };
   |                    ^^^^^^^^

error: this closing tag has no corresponding opening tag
  --> $DIR/html-element-fail.rs:13:18
   |
13 |     html! { <div></span></div> };
   |                  ^^^^^^^

error: only one root html element is allowed (hint: you can wrap multiple html elements in a fragment `<></>`)
  --> $DIR/html-element-fail.rs:14:20
   |
14 |     html! { <img /></img> };
   |                    ^^^^^^

error: expected a valid html element
  --> $DIR/html-element-fail.rs:15:18
   |
15 |     html! { <div>Invalid</div> };
   |                  ^^^^^^^

error: `attr` can only be specified once but is given here again
  --> $DIR/html-element-fail.rs:17:27
   |
17 |     html! { <input attr=1 attr=2 /> };
   |                           ^^^^

error: `value` can only be specified once but is given here again
  --> $DIR/html-element-fail.rs:18:32
   |
18 |     html! { <input value="123" value="456" /> };
   |                                ^^^^^

error: `kind` can only be specified once but is given here again
  --> $DIR/html-element-fail.rs:19:36
   |
19 |     html! { <input kind="checkbox" kind="submit" /> };
   |                                    ^^^^

error: `checked` can only be specified once but is given here again
  --> $DIR/html-element-fail.rs:20:33
   |
20 |     html! { <input checked=true checked=false /> };
   |                                 ^^^^^^^

error: `disabled` can only be specified once but is given here again
  --> $DIR/html-element-fail.rs:21:34
   |
21 |     html! { <input disabled=true disabled=false /> };
   |                                  ^^^^^^^^

error: `selected` can only be specified once but is given here again
  --> $DIR/html-element-fail.rs:22:35
   |
22 |     html! { <option selected=true selected=false /> };
   |                                   ^^^^^^^^

error: `class` can only be specified once but is given here again
  --> $DIR/html-element-fail.rs:23:32
   |
23 |     html! { <div class="first" class="second" /> };
   |                                ^^^^^

error: `ref` can only be specified once
  --> $DIR/html-element-fail.rs:38:20
   |
38 |     html! { <input ref=() ref=() /> };
   |                    ^^^

error: the tag `<input>` is a void element and cannot have children (hint: rewrite this as `<input/>`)
  --> $DIR/html-element-fail.rs:40:13
   |
40 |     html! { <input type="text"></input> };
   |             ^^^^^^^^^^^^^^^^^^^

error: the tag `<iNpUt>` is a void element and cannot have children (hint: rewrite this as `<iNpUt/>`)
  --> $DIR/html-element-fail.rs:41:13
   |
41 |     html! { <iNpUt type="text"></iNpUt> };
   |             ^^^^^^^^^^^^^^^^^^^

error: this dynamic tag is missing an expression block defining its value
  --> $DIR/html-element-fail.rs:43:14
   |
43 |     html! { <@></@> };
   |              ^

error: dynamic closing tags must not have a body (hint: replace it with just `</@>`)
  --> $DIR/html-element-fail.rs:44:27
   |
44 |     html! { <@{"test"}></@{"test"}> };
   |                           ^^^^^^^^

error: this dynamic tag is missing an expression block defining its value
  --> $DIR/html-element-fail.rs:46:14
   |
46 |     html! { <@/> };
   |              ^

error: boolean attributes don't support being used as an optional attribute (hint: a value of false results in the attribute not being set)
  --> $DIR/html-element-fail.rs:50:20
   |
50 |     html! { <input disabled?=Some(true) /> };
   |                    ^^^^^^^^

error: `checked` does not support being used as an optional attribute
  --> $DIR/html-element-fail.rs:58:20
   |
58 |     html! { <input checked?=Some(false) /> };
   |                    ^^^^^^^^^

error: `ref` does not support being used as an optional attribute
  --> $DIR/html-element-fail.rs:59:20
   |
59 |     html! { <input ref?=() /> };
   |                    ^^^^^

warning: use of deprecated item 'compile_fail::deprecated_use_of_class': the use of `(...)` with the attribute `class` is deprecated and will be removed in version 0.19. Use the `classes!` macro instead.
  --> $DIR/html-element-fail.rs:63:24
   |
63 |     html! { <div class=("deprecated", "warning") /> };
   |                        ^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(deprecated)]` on by default

error[E0308]: mismatched types
  --> $DIR/html-element-fail.rs:25:28
   |
25 |     html! { <input checked=1 /> };
   |                            ^ expected `bool`, found integer

error[E0308]: mismatched types
  --> $DIR/html-element-fail.rs:26:29
   |
26 |     html! { <input disabled=1 /> };
   |                             ^ expected `bool`, found integer

error[E0308]: mismatched types
  --> $DIR/html-element-fail.rs:27:30
   |
27 |     html! { <option selected=1 /> };
   |                              ^ expected `bool`, found integer

error[E0277]: `()` doesn't implement `std::fmt::Display`
  --> $DIR/html-element-fail.rs:28:25
   |
28 |     html! { <input type=() /> };
   |                         ^^ `()` cannot be formatted with the default formatter
   |
   = help: the trait `std::fmt::Display` is not implemented for `()`
   = note: in format strings you may be able to use `{:?}` (or {:#?} for pretty-print) instead
   = note: required because of the requirements on the impl of `std::string::ToString` for `()`
   = note: required by `std::string::ToString::to_string`

error[E0277]: `()` doesn't implement `std::fmt::Display`
  --> $DIR/html-element-fail.rs:29:26
   |
29 |     html! { <input value=() /> };
   |                          ^^ `()` cannot be formatted with the default formatter
   |
   = help: the trait `std::fmt::Display` is not implemented for `()`
   = note: in format strings you may be able to use `{:?}` (or {:#?} for pretty-print) instead
   = note: required because of the requirements on the impl of `std::string::ToString` for `()`

error[E0277]: `()` doesn't implement `std::fmt::Display`
  --> $DIR/html-element-fail.rs:30:21
   |
30 |     html! { <a href=() /> };
   |                     ^^ `()` cannot be formatted with the default formatter
   |
   = help: the trait `std::fmt::Display` is not implemented for `()`
   = note: in format strings you may be able to use `{:?}` (or {:#?} for pretty-print) instead
   = note: required because of the requirements on the impl of `std::string::ToString` for `()`
   = note: required by `std::string::ToString::to_string`

error[E0308]: mismatched types
  --> $DIR/html-element-fail.rs:32:28
   |
32 |     html! { <input onclick=1 /> };
   |                            ^ expected enum `yew::callback::Callback`, found integer
   |
   = note: expected enum `yew::callback::Callback<web_sys::features::gen_MouseEvent::MouseEvent>`
              found type `{integer}`

error[E0308]: mismatched types
  --> $DIR/html-element-fail.rs:33:28
   |
33 |     html! { <input onclick=Callback::from(|a: String| ()) /> };
   |                            ^^^^^^^^ expected struct `web_sys::features::gen_MouseEvent::MouseEvent`, found struct `std::string::String`
   |
   = note: expected enum `yew::callback::Callback<web_sys::features::gen_MouseEvent::MouseEvent>`
              found enum `yew::callback::Callback<std::string::String>`

error[E0277]: `NotToString` doesn't implement `std::fmt::Display`
  --> $DIR/html-element-fail.rs:35:27
   |
35 |     html! { <input string=NotToString /> };
   |                           ^^^^^^^^^^^ `NotToString` cannot be formatted with the default formatter
   |
   = help: the trait `std::fmt::Display` is not implemented for `NotToString`
   = note: in format strings you may be able to use `{:?}` (or {:#?} for pretty-print) instead
   = note: required because of the requirements on the impl of `std::string::ToString` for `NotToString`
   = note: required by `std::string::ToString::to_string`

error[E0308]: mismatched types
  --> $DIR/html-element-fail.rs:37:24
   |
37 |     html! { <input ref=() /> };
   |                        ^^ expected struct `yew::html::NodeRef`, found `()`

error[E0277]: the trait bound `std::borrow::Cow<'static, str>: std::convert::From<{integer}>` is not satisfied
  --> $DIR/html-element-fail.rs:45:15
   |
45 |     html! { <@{55}></@> };
   |               ^^^^ the trait `std::convert::From<{integer}>` is not implemented for `std::borrow::Cow<'static, str>`
   |
   = help: the following implementations were found:
             <std::borrow::Cow<'a, [T]> as std::convert::From<&'a [T]>>
             <std::borrow::Cow<'a, [T]> as std::convert::From<&'a std::vec::Vec<T>>>
             <std::borrow::Cow<'a, [T]> as std::convert::From<std::vec::Vec<T>>>
             <std::borrow::Cow<'a, std::ffi::CStr> as std::convert::From<&'a std::ffi::CStr>>
           and 11 others
   = note: required because of the requirements on the impl of `std::convert::Into<std::borrow::Cow<'static, str>>` for `{integer}`
   = note: required by `std::convert::Into::into`

error[E0308]: mismatched types
  --> $DIR/html-element-fail.rs:48:23
   |
48 |     html! { <a media?="media" /> };
   |                       ^^^^^^^
   |                       |
   |                       expected enum `std::option::Option`, found `&str`
   |                       help: try using a variant of the expected enum: `Some("media")`
   |
   = note:   expected enum `std::option::Option<_>`
           found reference `&'static str`

error[E0277]: `NotToString` doesn't implement `std::fmt::Display`
  --> $DIR/html-element-fail.rs:49:23
   |
49 |     html! { <a media?=Some(NotToString) /> };
   |                       ^^^^ `NotToString` cannot be formatted with the default formatter
   |
   = help: the trait `std::fmt::Display` is not implemented for `NotToString`
   = note: in format strings you may be able to use `{:?}` (or {:#?} for pretty-print) instead
   = note: required because of the requirements on the impl of `std::string::ToString` for `NotToString`
   = note: required by `std::string::ToString::to_string`

error[E0308]: mismatched types
  --> $DIR/html-element-fail.rs:51:26
   |
51 |     html! { <input type?="kind" /> };
   |                          ^^^^^^
   |                          |
   |                          expected enum `std::option::Option`, found `&str`
   |                          help: try using a variant of the expected enum: `Some("kind")`
   |
   = note:   expected enum `std::option::Option<_>`
           found reference `&'static str`

error[E0277]: `NotToString` doesn't implement `std::fmt::Display`
  --> $DIR/html-element-fail.rs:52:26
   |
52 |     html! { <input type?=Some(NotToString) /> };
   |                          ^^^^ `NotToString` cannot be formatted with the default formatter
   |
   = help: the trait `std::fmt::Display` is not implemented for `NotToString`
   = note: in format strings you may be able to use `{:?}` (or {:#?} for pretty-print) instead
   = note: required because of the requirements on the impl of `std::string::ToString` for `NotToString`
   = note: required by `std::string::ToString::to_string`

error[E0308]: mismatched types
  --> $DIR/html-element-fail.rs:53:24
   |
53 |     html! { <li value?="value" /> };
   |                        ^^^^^^^
   |                        |
   |                        expected enum `std::option::Option`, found `&str`
   |                        help: try using a variant of the expected enum: `Some("value")`
   |
   = note:   expected enum `std::option::Option<_>`
           found reference `&'static str`

error[E0277]: `NotToString` doesn't implement `std::fmt::Display`
  --> $DIR/html-element-fail.rs:54:24
   |
54 |     html! { <li value?=Some(NotToString) /> };
   |                        ^^^^ `NotToString` cannot be formatted with the default formatter
   |
   = help: the trait `std::fmt::Display` is not implemented for `NotToString`
   = note: in format strings you may be able to use `{:?}` (or {:#?} for pretty-print) instead
   = note: required because of the requirements on the impl of `std::string::ToString` for `NotToString`
   = note: required by `std::string::ToString::to_string`

error[E0308]: mismatched types
  --> $DIR/html-element-fail.rs:56:22
   |
56 |     html! { <a href?="href" /> };
   |                      ^^^^^^
   |                      |
   |                      expected enum `std::option::Option`, found `&str`
   |                      help: try using a variant of the expected enum: `Some("href")`
   |
   = note:   expected enum `std::option::Option<_>`
           found reference `&'static str`

error[E0277]: `NotToString` doesn't implement `std::fmt::Display`
  --> $DIR/html-element-fail.rs:57:22
   |
57 |     html! { <a href?=Some(NotToString) /> };
   |                      ^^^^ `NotToString` cannot be formatted with the default formatter
   |
   = help: the trait `std::fmt::Display` is not implemented for `NotToString`
   = note: in format strings you may be able to use `{:?}` (or {:#?} for pretty-print) instead
   = note: required because of the requirements on the impl of `std::string::ToString` for `NotToString`
   = note: required by `std::string::ToString::to_string`

error[E0308]: mismatched types
  --> $DIR/html-element-fail.rs:60:20
   |
60 |     html! { <input onfocus?=Some(5) /> };
   |                    ^^^^^^^ expected enum `yew::callback::Callback`, found integer
   |
   = note: expected enum `yew::callback::Callback<web_sys::features::gen_FocusEvent::FocusEvent>`
              found type `{integer}`

error[E0308]: mismatched types
  --> $DIR/html-element-fail.rs:61:29
   |
61 |     html! { <input onfocus?=Callback::from(|_| ()) /> };
   |                             ^^^^^^^^^^^^^^^^^^^^^^
   |                             |
   |                             expected enum `std::option::Option`, found enum `yew::callback::Callback`
   |                             help: try using a variant of the expected enum: `Some(Callback::from(|_| ()))`
   |
   = note: expected enum `std::option::Option<_>`
              found enum `yew::callback::Callback<_>`
